src/nvim/edit.c

```diff
@@ -5281,10 +5283,16 @@ insertchar (
 
       (*mb_char2bytes)(c, buf);
       buf[cc] = NUL;
-      ins_char_bytes(buf, cc, false);
+      ins_char_bytes(buf, cc);
+      // TODO(timeyyy): test this code path...
+      extmark_col_adjust(curbuf, curwin->w_cursor.col, curwin->w_cursor.col, 0,
+                         (long)STRLEN(cc), kextmarkNoReverse);
       AppendCharToRedobuff(c);
     } else {
-      ins_char(c, true);
+      ins_char(c);
+      char_u buf[MB_MAXBYTES + 1];
+      extmark_col_adjust(curbuf, curwin->w_cursor.col, curwin->w_cursor.col, 0,
+                         (long)(*mb_char2bytes)(c, buf), kextmarkNoReverse);
```

Comment:
or maybe the upper branch should just be removed, `ins_char` is perfectly capable of handle multibyte characters. Then it should return the number of bytes inserted, so we don't need to call `mb_char2bytes` twice.